### PR TITLE
nfd-master: implement --instance flag

### PR DESCRIPTION
**Pull request by marquiz**
This can be used to help running multiple parallel NFD deployments in
the same cluster. The flag changes the node annotation namespace to
<instance>.nfd.node.kubernetes.io allowing different nfd-master intances
to store metadata in separate annotations.

Related to #427

### CODE REVIEW

**Comment by e0ne** [comment]
It doesn't fixes #427. Adding new annotations works well, but different NDF instances will override each other labels. It means any software which relates on standard NFD behaviour and labels won't work good in case when two or more NFD are deployed.

**Comment by marquiz** [comment]
> It doesn't fixes #427. Adding new annotations works well, but different NDF instances will override each other labels. It means any software which relates on standard NFD behaviour and labels won't work good in case when two or more NFD are deployed.

My point is that you need to configure the "label separation" somewhere, either master or worker side. Does it really matter if it would be on worker side? Worker side is much more flexible, e.g. you can directly disable all unneeded feature sources.

**Comment by e0ne** [comment]
@marquiz is it a supported configuration to have 2+ workers on the same node with different configuration?

**Comment by marquiz** [comment]
> override each other labels

What kind of label overwrite are you concerned about? If two instances write cpuid.XYZ or kernel version it shouldn't even matter as the value from both sources should be the same. When running multiple instances of NFD in parallel there is really no reliable way prevent overlap, i.e. prevent different instances modifying the same labels. An incorrect deployment/configuration of one NFD instance will ruin everything.

**Comment by marquiz** [comment]
> @marquiz is it a supported configuration to have 2+ workers on the same node with different configuration?

Yes, sure. They are separate instances (pods) so they may have different configuration. The "configuration" is basically the command line flags plus the configuration file. Configuration file is typically taken from a config map. Just use different configmap for different instances (e.g. run them in separate k8s namespaces)

**Comment by e0ne** [comment]
E..g.:
instance1: --label-whitelist=pci-0300_102b.present"
instance2: --label-whitelist=pci-8086.sriov.capable"

As a result, I would like to have a list of labels which contains both pci-0300_102b and pci-8086.sriov features discovered

**Comment by e0ne** [comment]
In other words, master should merge labels provided by different workers

**Comment by zvonkok** [comment]
@e0ne Why would different workers have different labels? Two instances of NFD would be running in e.g. different namespaces, not on different nodes.

**Comment by zvonkok** [comment]
Each worker would discover the very same features. @e0ne Are you planning to run the second instance of NFD on a subset of available nodes? Usually NFD is deployed to all workers.

**Comment by e0ne** [comment]
@zvonkok here is an example how we deploy NFD: https://github.com/kubernetes-sigs/node-feature-discovery/issues/427#issuecomment-772554744

**Comment by zvonkok** [comment]
Ah got it. You have different configurations for each instance of NFD.

**Comment by e0ne** [comment]
> Ah got it. You have different configurations for each instance of NFD.

Exactly, that's why we need several instances of NDF

**Comment by marquiz** [comment]
> In other words, master should merge labels provided by different workers

As I commented in #427, this totally different and orthogonal AFAICS. Could you open a separate issue about this where we could discuss it?

**Comment by e0ne** [comment]
we can add 'instance' prefix to labels, not for annotations and it should work in a such way: each NFD instance will override labels with 'instance' prefix and can deploy a worker without prefix as a default

**Comment by marquiz** [comment]
> we can add 'instance' prefix to labels, not for annotations and it should work in a such way: each NFD instance will override labels with 'instance' prefix and can deploy a worker without prefix as a default

This sounds like not a very good idea and would just add confusion, I think.

I still fail to see what you'd gain with this. If two deployments would create the same label (name) with a different value this is clearly a configuration error. They should use a different label name, e.g. by using a dedicated label ns like `vendor-1.com/feature-x` and `vendor-2.com/feature-x`. If you're talking about the "built-in" feature labels like `cpu-cpuid.XYZ` or `pci-<device>.present` etc the value really should not be different between deployments

**Review by moshe010 — APPROVED** [review]
Thanks for working on this. We tested it in it solves our problem :)

**Comment by marquiz** [comment]
> Thanks for working on this. We tested it in it solves our problem :)

Thanks for testing. I haven't found any problems in my own test setup. Just waiting for a review....

What I really want to have is an e2e-test for this but I think that could be addressed in a separate PR

**Comment by moshe010** [comment]
@zvonkok, Can you please review this PR?

**Comment by marquiz** [comment]
@moshe010 you can also take a look at the changes and provide comments, that would be much appreciated 😉

File: pkg/nfd-master/nfd-master.go
Comment (adrianchiris) [inline pkg/nfd-master/nfd-master.go:48]: nit: base Namespace ?

File: cmd/nfd-master/main.go
Comment (e0ne) [inline cmd/nfd-master/main.go:117]: It would be good to add some help message here or in line 80 and provide details what instance name format is expected

File: pkg/nfd-master/nfd-master.go
Comment (e0ne) [inline pkg/nfd-master/nfd-master.go:75]: nit: do we want to keep all items in an alphabetical order?

File: pkg/nfd-master/nfd-master.go
Comment (adrianchiris) [inline pkg/nfd-master/nfd-master.go:114]: nit:
should you use strings.join() like in other places in this file ?

File: pkg/nfd-master/nfd-master.go
Comment (adrianchiris) [inline pkg/nfd-master/nfd-master.go:518]: ran this on windows ?

nit: not directly related to this commit

File: test/e2e/node_feature_discovery.go
Comment (adrianchiris) [inline test/e2e/node_feature_discovery.go:396]: in case instance is used it will break cleanup right ?

if/when support is added to instance flag in e2e test this would need to be modified.

**Review by adrianchiris — COMMENTED** [review]
Changes look good IMO,

first commits are refactors which are not necessarily needed for support for instance flag, but are sensible and simplifies code.

just a few very minor nits from me plus a typo in cmdline reference.

**Comment by marquiz** [comment]
Thanks for the review @e0ne and @adrianchiris ! Addressed the comments in the topmost patch

I want to squash the fixes in the other patches before merging
/hold

File: test/e2e/node_feature_discovery.go
Comment (marquiz) [reply test/e2e/node_feature_discovery.go:396]: Yes, correct. My plan was to update/extend the e2e tests in a separate PR

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [reply pkg/nfd-master/nfd-master.go]: > nit: not directly related to this commit

True. Reverted this change here and addressed in a separate PR (#435)

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [reply pkg/nfd-master/nfd-master.go:115]: The other places have a slice to join. I originally used strings.Join but decided chaged it. Here we only have two values so IMHO this is more readable and practical than doing
```
strings.Join([args.Instance, AnnotationNsBase], ".")
```

WDYT?

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [reply pkg/nfd-master/nfd-master.go]: Yes, fixed

File: docs/advanced/master-commandline-reference.md
Comment (marquiz) [reply docs/advanced/master-commandline-reference.md]: Thanks, fixed

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [reply pkg/nfd-master/nfd-master.go]: Yes, certainly, thanks for noting this. Fixed

File: cmd/nfd-master/main.go
Comment (marquiz) [reply cmd/nfd-master/main.go]: Indeed 😄  Added an explanation to the error message

**Comment by marquiz** [comment]
> first commits are refactors which are not necessarily needed for support for instance flag, but are sensible and simplifies code.

True, the refactors are not strictly necessary, but without them the code would far uglier as the annotation ns mangling messes things up. I tried and didn't like it 😄  I didn't see much point in putting the refactors in a separate PR, but, could be easily done if people like it better that way

**Comment by adrianchiris** [comment]
> True, the refactors are not strictly necessary, but without them the code would far uglier as the annotation ns mangling messes things up. I tried and didn't like it 😄 I didn't see much point in putting the refactors in a separate PR, but, could be easily done if people like it better that way

Im totally fine with having in the same PR.

File: pkg/nfd-master/nfd-master.go
Comment (adrianchiris) [reply pkg/nfd-master/nfd-master.go:115]: Make sense, just wanted to point it out and i see you already considered it when coding.

**Review by adrianchiris — APPROVED** [review]
Thanks for addressing my comments @marquiz  !

**Comment by marquiz** [comment]
Needed to rebase. Squashed the review fixes in the patchset.
/unhold

**Comment by e0ne** [comment]
@marquiz thanks for addressing my comments!

**Pull request merged**
